### PR TITLE
fix(trimObject): Ignore __proto__ keys that are passed in

### PR DESCRIPTION
**Pull request by alexvision**
This protects against objects that have been parsed with `JSON.parse` which will include `__proto__` as a key.

Relatively low risk vector, but better to be safe than sorry I guess.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -10,7 +10,8 @@ module.exports = {
     var out = {};
     for (var i = 0; i < keys.length; i++) {
       var val = obj[keys[i]];
-      if (val === '' || val === null || val === undefined) continue;
+      // Ignore __proto__ to avoid prototype pollution
+      if (val === '' || val === null || val === undefined || val === '__proto__') continue;
```
Comment (pxwise) [inline index.js]: I'd separate `__proto__` from this condition, separate concerns.

We should also be testing for all combos of ASCII and unicode escape. We can borrow this regex: https://github.com/hapijs/bourne/blob/master/lib/index.js#L5

Of course adding that may have perf implications. Is this used in hot code paths? may be worth benchmarking the perf of this function.

**Comment by STRML** [comment]
You also don't want to test the value, you want to test the key.

On Tue, Jun 25, 2019, 2:59 PM Shawn Stedman <notifications@github.com>
wrote:

> *@pxwise* commented on this pull request.
> ------------------------------
>
> In index.js
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_BitMEX_bitmex-2Dshared_pull_1-23discussion-5Fr297367585&d=DwMCaQ&c=o2Rg2hCP31Es1DdrOYH75Q&r=GQjFTkW-64PxBigEJWtmV9xmwYeJF6OQpE-3EKPViq0&m=u0bKQSGO38IcRqc_ePRP8ZhY0yTX9aIAMojCynyELzo&s=2GPdOEQM4ZUOY-kYvmDa6IVJutxBElzN-qI9r_LWZSA&e=>
> :
>
> > @@ -10,7 +10,8 @@ module.exports = {
>      var out = {};
>      for (var i = 0; i < keys.length; i++) {
>        var val = obj[keys[i]];
> -      if (val === '' || val === null || val === undefined) continue;
> +      // Ignore __proto__ to avoid prototype pollution
> +      if (val === '' || val === null || val === undefined || val === '__proto__') continue;
>
> I'd separate __proto__ from this condition, separate concerns.
>
> We should also be testing for all combos of ASCII and unicode escape. We
> can borrow this regex:
> https://github.com/hapijs/bourne/blob/master/lib/index.js#L5
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_hapijs_bourne_blob_master_lib_index.js-23L5&d=DwMCaQ&c=o2Rg2hCP31Es1DdrOYH75Q&r=GQjFTkW-64PxBigEJWtmV9xmwYeJF6OQpE-3EKPViq0&m=u0bKQSGO38IcRqc_ePRP8ZhY0yTX9aIAMojCynyELzo&s=t1kgjIhu5xZ42ErPqSEAF2mst7DUidTY33_ym1fRXIk&e=>
>
> Of course adding that may have perf implications. Is this used in hot code
> paths? may be worth benchmarking the perf of this function.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_BitMEX_bitmex-2Dshared_pull_1-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAAJEKP3Z7GWEJVPWXO2V6RTP4J2J3A5CNFSM4H3CJ7YKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4TVOFI-23pullrequestreview-2D254236437&d=DwMCaQ&c=o2Rg2hCP31Es1DdrOYH75Q&r=GQjFTkW-64PxBigEJWtmV9xmwYeJF6OQpE-3EKPViq0&m=u0bKQSGO38IcRqc_ePRP8ZhY0yTX9aIAMojCynyELzo&s=P_Nwmgu1_KH0WQBxfvKo7jTE8HM1i05ky8377B4SVwE&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAJEKP7KYYTIUMVQMXNFHILP4J2J3ANCNFSM4H3CJ7YA&d=DwMCaQ&c=o2Rg2hCP31Es1DdrOYH75Q&r=GQjFTkW-64PxBigEJWtmV9xmwYeJF6OQpE-3EKPViq0&m=u0bKQSGO38IcRqc_ePRP8ZhY0yTX9aIAMojCynyELzo&s=hsWQ2Z9Uh9_vrJIFLpj-WDW9s5iTj1oJxyoPI-OFr5g&e=>
> .
>

-- 


This e-mail and all information in, attached to, or linked via this 
e-mail (together the ‘e-mail’) is confidential and may be legally 
privileged. It is intended solely for the intended addressee(s). Access to, 
or any onward transmission, of this e-mail by any other person is not 
authorised. If you are not the intended recipient, you are requested to 
immediately alert the sender of this e-mail and to immediately delete this 
e-mail. Any disclosure in any form of all or part of this e-mail, or of any 
the parties to it, including any copying, distribution or any action taken 
or omitted to be taken in reliance on it, is prohibited and may be 
unlawful. 




Transacting on BitMEX is not permitted for, or offered or 
available to, any person(s) located in or being resident(s) of (i) the 
United States of America or Québec (Canada). Additionally, transacting on 
BitMEX is not permitted for, or offered or available to, any person(s) 
located in or being resident(s) of (i) Cuba, Crimea and Sevastopol, Iran, 
Syria, North Korea, or Sudan, (ii) any other sanctioned jurisdiction, or 
(iii) any jurisdiction where the services offered by BitMEX are restricted. 





This e-mail is not, and is not intended to be, and should not be 
construed as being, (a) any offer, solicitation, or promotion of any kind; 
(b) the basis of any investment or other decision(s);  (c) any 
recommendation to buy, sell or transact in any manner any good(s), 
product(s) or service(s), nor engage in any investment(s) or other 
transaction(s) or activities;  or (d) the provision of, or related to, any 
advisory service(s) or activities, including regarding any investment, tax, 
legal, financial, accounting, consulting or any other related service(s).

-- 


This e-mail and all information in, attached to, or linked via this 
e-mail (together the ‘e-mail’) is confidential and may be legally 
privileged. It is intended solely for the intended addressee(s). Access to, 
or any onward transmission, of this e-mail by any other person is not 
authorised. If you are not the intended recipient, you are requested to 
immediately alert the sender of this e-mail and to immediately delete this 
e-mail. Any disclosure in any form of all or part of this e-mail, or of any 
the parties to it, including any copying, distribution or any action taken 
or omitted to be taken in reliance on it, is prohibited and may be 
unlawful. 




Transacting on BitMEX is not permitted for, or offered or 
available to, any person(s) located in or being resident(s) of (i) the 
United States of America or Québec (Canada). Additionally, transacting on 
BitMEX is not permitted for, or offered or available to, any person(s) 
located in or being resident(s) of (i) Cuba, Crimea and Sevastopol, Iran, 
Syria, North Korea, or Sudan, (ii) any other sanctioned jurisdiction, or 
(iii) any jurisdiction where the services offered by BitMEX are restricted. 





This e-mail is not, and is not intended to be, and should not be 
construed as being, (a) any offer, solicitation, or promotion of any kind; 
(b) the basis of any investment or other decision(s);  (c) any 
recommendation to buy, sell or transact in any manner any good(s), 
product(s) or service(s), nor engage in any investment(s) or other 
transaction(s) or activities;  or (d) the provision of, or related to, any 
advisory service(s) or activities, including regarding any investment, tax, 
legal, financial, accounting, consulting or any other related service(s).

**Comment by jhob** [comment]
^^^ and unit tests would've caught that easily

**Comment by alexvision** [comment]
Whoops yeah, stupid mistake, pushed an old version of this. I'll take the time to include unit tests on these to make sure it is doing what it says on the box

File: index.js (hunk 1 above)
Comment (sandbochs) [reply index.js]: Wouldn't the unicode escape appear as as underscores already at this point?

**Comment by alexvision** [comment]
Fixed the test/checking wrong value issue. Adding either Jest or Mocha really blows out the dependencies, but given it is just a dev dependency it should be fine. I'm looking into the unicode stuff now, I'm not super familiar with it so I can try and add the regex if needed, but if we don't need it per Elliot's comment this should work as is

**Comment by STRML** [comment]
Definitely won't need unicode escaping here, setting an object property
does not automatically unescape!

On Tue, Jun 25, 2019, 4:44 PM Alex Manusu <notifications@github.com> wrote:

> Fixed the test/checking wrong value issue. Adding either Jest or Mocha
> really blows out the dependencies, but given it is just a dev dependency it
> should be fine. I'm looking into the unicode stuff now, I'm not super
> familiar with it so I can try and add the regex if needed, but if we don't
> need it per Elliot's comment this should work as is
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_BitMEX_bitmex-2Dshared_pull_1-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAAJEKP4GUJIVLPCHEWGD7GLP4KGSDA5CNFSM4H3CJ7YKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYRV33Y-23issuecomment-2D505634287&d=DwMCaQ&c=o2Rg2hCP31Es1DdrOYH75Q&r=GQjFTkW-64PxBigEJWtmV9xmwYeJF6OQpE-3EKPViq0&m=Y03T6rGA2fu1HFrBwG-mgn8FB_AvELIhNXq9sJFMmdc&s=XT2hq6ijhRBwh4flMn2-5XfYYh1D37lGmJW7_7vWCVQ&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAJEKP5Q743YYMQZAZRVEU3P4KGSDANCNFSM4H3CJ7YA&d=DwMCaQ&c=o2Rg2hCP31Es1DdrOYH75Q&r=GQjFTkW-64PxBigEJWtmV9xmwYeJF6OQpE-3EKPViq0&m=Y03T6rGA2fu1HFrBwG-mgn8FB_AvELIhNXq9sJFMmdc&s=-ig6MCemNKH8KNAzocsSAHcRCl367eo-P6COxNHrXzI&e=>
> .
>

-- 


This e-mail and all information in, attached to, or linked via this 
e-mail (together the ‘e-mail’) is confidential and may be legally 
privileged. It is intended solely for the intended addressee(s). Access to, 
or any onward transmission, of this e-mail by any other person is not 
authorised. If you are not the intended recipient, you are requested to 
immediately alert the sender of this e-mail and to immediately delete this 
e-mail. Any disclosure in any form of all or part of this e-mail, or of any 
the parties to it, including any copying, distribution or any action taken 
or omitted to be taken in reliance on it, is prohibited and may be 
unlawful. 




Transacting on BitMEX is not permitted for, or offered or 
available to, any person(s) located in or being resident(s) of (i) the 
United States of America or Québec (Canada). Additionally, transacting on 
BitMEX is not permitted for, or offered or available to, any person(s) 
located in or being resident(s) of (i) Cuba, Crimea and Sevastopol, Iran, 
Syria, North Korea, or Sudan, (ii) any other sanctioned jurisdiction, or 
(iii) any jurisdiction where the services offered by BitMEX are restricted. 





This e-mail is not, and is not intended to be, and should not be 
construed as being, (a) any offer, solicitation, or promotion of any kind; 
(b) the basis of any investment or other decision(s);  (c) any 
recommendation to buy, sell or transact in any manner any good(s), 
product(s) or service(s), nor engage in any investment(s) or other 
transaction(s) or activities;  or (d) the provision of, or related to, any 
advisory service(s) or activities, including regarding any investment, tax, 
legal, financial, accounting, consulting or any other related service(s).

-- 


This e-mail and all information in, attached to, or linked via this 
e-mail (together the ‘e-mail’) is confidential and may be legally 
privileged. It is intended solely for the intended addressee(s). Access to, 
or any onward transmission, of this e-mail by any other person is not 
authorised. If you are not the intended recipient, you are requested to 
immediately alert the sender of this e-mail and to immediately delete this 
e-mail. Any disclosure in any form of all or part of this e-mail, or of any 
the parties to it, including any copying, distribution or any action taken 
or omitted to be taken in reliance on it, is prohibited and may be 
unlawful. 




Transacting on BitMEX is not permitted for, or offered or 
available to, any person(s) located in or being resident(s) of (i) the 
United States of America or Québec (Canada). Additionally, transacting on 
BitMEX is not permitted for, or offered or available to, any person(s) 
located in or being resident(s) of (i) Cuba, Crimea and Sevastopol, Iran, 
Syria, North Korea, or Sudan, (ii) any other sanctioned jurisdiction, or 
(iii) any jurisdiction where the services offered by BitMEX are restricted. 





This e-mail is not, and is not intended to be, and should not be 
construed as being, (a) any offer, solicitation, or promotion of any kind; 
(b) the basis of any investment or other decision(s);  (c) any 
recommendation to buy, sell or transact in any manner any good(s), 
product(s) or service(s), nor engage in any investment(s) or other 
transaction(s) or activities;  or (d) the provision of, or related to, any 
advisory service(s) or activities, including regarding any investment, tax, 
legal, financial, accounting, consulting or any other related service(s).

File: index.js (hunk 1 above)
Comment (pxwise) [reply index.js]: Ah, right. unicode is only a concern when the key is `JSON.parse`d
